dll/win32/shell32/droptargets/CFSDropTarget.cpp

```diff
@@ -547,104 +547,126 @@ HRESULT CFSDropTarget::_DoDrop(IDataObject *pDataObject,
-
-            TRACE("target path = %s", debugstr_w(wszTargetPath));
+            TRACE("target path = %s\n", debugstr_w(m_sPathTarget));
 
             /* We need to create a link for each pidl in the copied items, so step through the pidls from the clipboard */
             for (UINT i = 0; i < lpcida->cidl; i++)
             {
-                //Find out which file we're copying
+                // Find out which file we're linking.
                 STRRET strFile;
                 hr = psfFrom->GetDisplayNameOf(apidl[i], SHGDN_FORPARSING, &strFile);
-                if (FAILED(hr))
-                {
-                    ERR("Error source obtaining path");
+                if (FAILED_UNEXPECTEDLY(hr))
                     break;
-                }
 
                 hr = StrRetToBufW(&strFile, apidl[i], wszPath, _countof(wszPath));
-                if (FAILED(hr))
-                {
-                    ERR("Error putting source path into buffer");
+                if (FAILED_UNEXPECTEDLY(hr))
                     break;
-                }
-                TRACE("source path = %s", debugstr_w(wszPath));
 
-                // Creating a buffer to hold the combined path
-                WCHAR buffer_1[MAX_PATH] = L"";
-                WCHAR *lpStr1;
-                lpStr1 = buffer_1;
+                TRACE("source path = %s\n", debugstr_w(wszPath));
 
+                WCHAR wszDisplayName[MAX_PATH];
                 LPWSTR pwszFileName = PathFindFileNameW(wszPath);
-                LPWSTR pwszExt = PathFindExtensionW(wszPath);
-                LPWSTR placementPath = PathCombineW(lpStr1, m_sPathTarget, pwszFileName);
-                CComPtr<IPersistFile> ppf;
-
-                //Check to see if it's already a link.
-                if (!wcsicmp(pwszExt, L".lnk"))
+                if (PathIsRootW(wszPath)) // Drive?
                 {
-                    //It's a link so, we create a new one which copies the old.
-                    if(!_GetUniqueFileName(placementPath, pwszExt, wszTarget, TRUE))
-                    {
-                        ERR("Error getting unique file name");
-                        hr = E_FAIL;
+                    hr = psfFrom->GetDisplayNameOf(apidl[i], 0, &strFile);
+                    if (FAILED_UNEXPECTEDLY(hr))
                         break;
-                    }
-                    hr = IShellLink_ConstructFromPath(wszPath, IID_PPV_ARG(IPersistFile, &ppf));
-                    if (FAILED(hr)) {
-                        ERR("Error constructing link from file");
+
+                    hr = StrRetToBufW(&strFile, apidl[i], wszDisplayName, _countof(wszDisplayName));
+                    if (FAILED_UNEXPECTEDLY(hr))
                         break;
+
+                    // Delete a ':' in wszDisplayName.
+                    LPWSTR pch0 = wcschr(wszDisplayName, L':'), pch1;
+                    if (pch0)
+                    {
+                        pch1 = pch0;
```

Comment:
This and the for-loop looks like a `MoveMemory(pch0, pch0 + 1, (wcslen(pch0 + 1) + 1)* sizeof(WCHAR));` to me.

Reply:
Exactly.